File: sdk/ruby-holons/lib/holons/serve.rb
# frozen_string_literal: true

require "fileutils"
require "socket"
require "thread"
require "tmpdir"

require_relative "describe"
require_relative "observability"
require_relative "transport"

module Holons
  module Serve
    SIGNALS = %w[INT TERM].freeze
    DEFAULT_SIGNAL_WAIT_INTERVAL = 0.25
    ParsedFlags = Struct.new(:listen_uri, :reflect, keyword_init: true)
    MemberRef = Struct.new(:slug, :address, :uid, keyword_init: true)
    ServeOptions = Struct.new(:reflect, :member_endpoints, :slug, keyword_init: true) do
      def initialize(**opts)
        super
        self.reflect = false if reflect.nil?
        self.member_endpoints ||= []
        self.slug ||= ""
      end
    end

    class << self
      # Parse --listen or --port from args.
      def parse_flags(args)
        parse_options(args).listen_uri
      end

      def parse_options(args)
        listen_uri = Transport::DEFAULT_URI
        reflect = false

        args.each_with_index do |arg, i|
          listen_uri = args[i + 1] if arg == "--listen" && i + 1 < args.length
          listen_uri = "tcp://:#{args[i + 1]}" if arg == "--port" && i + 1 < args.length
          reflect = true if arg == "--reflect"
        end

        ParsedFlags.new(listen_uri: listen_uri, reflect: reflect)
      end

      def run(listen_uri, register)
        run_with_options(listen_uri, register, false)
      end

      def run_with_options(listen_uri, register, reflect = false, on_listen: nil, env: ENV)
        run_with_serve_options(
          listen_uri,
          register,
          ServeOptions.new(reflect: reflect),
          on_listen: on_listen,
          env: env
        )
      end

      def run_with_serve_options(listen_uri, register, options = ServeOptions.new, on_listen: nil, env: ENV)
        ensure_grpc_runtime!
        raise ArgumentError, "register callback is required" unless register.respond_to?(:call)

        options ||= ServeOptions.new
        parsed = Transport.parse_uri(listen_uri)
        server = GRPC::RpcServer.new
        register.call(server)
        auto_register_describe(server)
        observability = auto_register_observability(server, env, options)
        reflection_enabled = maybe_register_reflection(server, options.reflect)

        actual_uri, cleanup, stdio_bridge = prepare_runtime(parsed, server)
        mode = reflection_mode(reflection_enabled)
        prom_server = start_prom_server(observability)
        metrics_addr = prom_server&.start.to_s
        start_observability_runtime(observability, actual_uri, parsed.scheme, metrics_addr)
        started_relays = []

        begin
          with_signal_handlers(server) do
            run_server(server) do
              stdio_bridge&.connect_to_server
              publish_listen_uri(actual_uri, on_listen) unless parsed.scheme == "stdio"
              on_listen&.call(actual_uri) if parsed.scheme == "stdio"
              warn("gRPC server listening on #{actual_uri} (#{mode})")
              started_relays = start_member_relays(observability, options.member_endpoints)
            end
          end
        ensure
          started_relays.each do |relay, channel|
            relay.stop
            Holons::DiscoverySupport.close_channel(channel) if defined?(Holons::DiscoverySupport)
          end
          prom_server&.close
          stdio_bridge&.close
          cleanup&.call
        end
      end

      private

      def ensure_grpc_runtime!
        require "grpc"
      rescue LoadError => e
        raise e
      end

      def auto_register_describe(server)
        Describe.register(server)
      rescue Describe::ErrNoIncodeDescription
        raise
      rescue StandardError => e
        warn("HolonMeta registration failed: #{e.message}")
        raise
      end

      def auto_register_observability(server, env, options = ServeOptions.new)
        Observability.check_env(env)
        raw = (env["OP_OBS"] || "").strip
        return nil if raw.empty?

        inst = Observability.from_env(Observability::Config.new(slug: options.slug.to_s), env: env)
        return nil if inst.families.empty?

        Observability.register_grpc_service(server, inst)
        inst
      end

      def start_observability_runtime(inst, actual_uri, transport, metrics_addr = "")
        return if inst.nil?

        run_dir = inst.cfg.run_dir.to_s
        Observability.enable_disk_writers(run_dir) unless run_dir.empty?
        inst.emit(Observability::EVENT_TYPES[:instance_ready], "listener" => actual_uri) if inst.enabled?(:events)
        return if run_dir.empty?

        Observability.write_meta_json(
          run_dir,
          Observability::MetaJson.new(
            slug: inst.cfg.slug,
            uid: inst.cfg.instance_uid,
            pid: Process.pid,
            started_at: Time.now,
            transport: transport,
            address: actual_uri,
            metrics_addr: metrics_addr,
            log_path: inst.enabled?(:logs) ? File.join(run_dir, "stdout.log") : "",
            organism_uid: inst.cfg.organism_uid,
            organism_slug: inst.cfg.organism_slug
          )
        )
      rescue StandardError
        nil
      end

      def start_prom_server(inst)
        return nil if inst.nil? || !inst.enabled?(:prom)

        server = Observability::PromServer.new(inst.cfg.prom_addr.to_s.empty? ? ":0" : inst.cfg.prom_addr)
        server.start
        server
      rescue StandardError => e
        warn("warning: prom HTTP bind failed: #{e.message}")
        nil
      end

      def start_member_relays(inst, members)
        return [] if inst.nil? || (!inst.enabled?(:logs) && !inst.enabled?(:events))

        Observability.require_grpc_observability_support!
        Array(members).each_with_object([]) do |raw, started|
          member = normalize_member_ref(raw)
          if member.slug.to_s.empty? || member.address.to_s.empty?
            warn(%(warning: observability relay skipped incomplete member ref: slug="#{member.slug}" uid="#{member.uid}" address="#{member.address}"))
            next
          end

          channel = nil
          begin
            target = normalize_relay_dial_target(member.address)
            channel = Holons::DiscoverySupport.dial_ready(target, 5000)
            member = resolve_relay_member_identity(channel, member)
            if member.uid.to_s.empty?
              warn("warning: observability relay uid unresolved for #{member.slug} at #{member.address}; chain hops will have empty uid")
            end
            stub = Holons::V1::HolonObservability::Stub.new(
              "unused",
              :this_channel_is_insecure,
              channel_override: channel
            )
            relay = Observability::MemberRelay.new(
              child_slug: member.slug,
              child_uid: member.uid,
              stub: stub,
              observability: inst
            )
            relay.start
            started << [relay, channel]
            channel = nil
          rescue StandardError => e
            Holons::DiscoverySupport.close_channel(channel) unless channel.nil?
            warn("warning: observability relay start #{member.slug}/#{member.uid}: #{e.message}")
          end
        end
      end

      def normalize_member_ref(raw)
        if raw.respond_to?(:slug) && raw.respond_to?(:address)
          return MemberRef.new(
            slug: raw.slug.to_s.strip,
            address: raw.address.to_s.strip,
            uid: raw.respond_to?(:uid) ? raw.uid.to_s.strip : ""
          )
        end

        MemberRef.new(slug: "", address: "", uid: "")
      end

      def resolve_relay_member_identity(channel, member)
        return member unless member.uid.to_s.empty?

        stub = Holons::V1::HolonObservability::Stub.new(
          "unused",
          :this_channel_is_insecure,
          channel_override: channel,
          timeout: 2
        )
        begin
          request = Holons::V1::EventsRequest.new(types: [:INSTANCE_READY], follow: false)
          stub.events(request).each do |event|
            next if event.instance_uid.to_s.empty? || !event.chain.empty?

            return MemberRef.new(
              slug: event.slug.to_s.empty? ? member.slug : event.slug.to_s,
              uid: event.instance_uid.to_s,
              address: member.address
            )
          end
        rescue StandardError
          nil
        end

        begin
          snapshot = stub.metrics(Holons::V1::MetricsRequest.new)
          return member if snapshot.instance_uid.to_s.empty?

          MemberRef.new(
            slug: snapshot.slug.to_s.empty? ? member.slug : snapshot.slug.to_s,
            uid: snapshot.instance_uid.to_s,
            address: member.address
          )
        rescue StandardError
          member
        end
      end

      def normalize_relay_dial_target(address)
        Holons::DiscoverySupport.normalize_dial_target(address)
      rescue StandardError
        address.to_s
      end


      def prepare_runtime(parsed, server)
        case parsed.scheme
        when "tcp"
          actual_uri = bind_tcp(server, parsed)
          [actual_uri, nil, nil]
        when "unix"
          cleanup, actual_uri = bind_unix(server, parsed.path)
          [actual_uri, cleanup, nil]
        when "stdio"
          bridge = StdioServeBridge.new(server: server)
          socket_path = bridge.start
          bound = server.add_http2_port("unix:#{socket_path}", :this_port_is_insecure)
          raise "failed to bind stdio bridge on #{socket_path}" if bound.to_i <= 0

          ["stdio://", nil, bridge]
        else
          raise ArgumentError, "unsupported transport URI: #{parsed.raw}"
        end
      end

      def bind_tcp(server, parsed)
        bind_host = parsed.host.to_s.strip
        bind_host = "0.0.0.0" if bind_host.empty?
        bind_port = parsed.port || 9090

        actual_port = server.add_http2_port("#{bind_host}:#{bind_port}", :this_port_is_insecure)
        raise "failed to bind tcp://#{bind_host}:#{bind_port}" if actual_port.to_i <= 0

        "tcp://#{normalize_host(bind_host)}:#{actual_port}"
      end

      def bind_unix(server, path)
        raise ArgumentError, "unix socket path is required" if path.to_s.strip.empty?

        FileUtils.mkdir_p(File.dirname(path))
        File.delete(path) if File.exist?(path)

        bound = server.add_http2_port("unix:#{path}", :this_port_is_insecure)
        raise "failed to bind unix://#{path}" if bound.to_i <= 0

        [proc { File.delete(path) if File.exist?(path) }, "unix://#{path}"]
      end

      def normalize_host(host)
        case host
        when "", "0.0.0.0", "::", "[::]"
          "127.0.0.1"
        else
          host
        end
      end

      def reflection_mode(reflect)
        return "reflection OFF" unless reflect

        "reflection ON"
      end

      def maybe_register_reflection(server, reflect)
        return false unless reflect

        require_relative "../grpc_reflection"
        server.handle(GrpcReflection::Server)
        server.handle(GrpcReflection::ServerAlpha)
        true
      end

      def publish_listen_uri(actual_uri, on_listen)
        if on_listen
          on_listen.call(actual_uri)
          return
        end

        puts(actual_uri)
        $stdout.flush if $stdout.respond_to?(:flush)
      end

      def run_server(server)
        server_error = nil
        server_thread = Thread.new do
          begin
            server.run_till_terminated
          rescue StandardError => e
            server_error = e
          end
        end

        begin
          raise "gRPC server failed to start" unless server.wait_till_running(5)
          yield if block_given?
          server_thread.join
        ensure
          stop_server(server)
          server_thread.join(0.2) if server_thread&.alive?
        end

        raise server_error unless server_error.nil?
      end

      def with_signal_handlers(server)
        previous = {}
        SIGNALS.each do |signal_name|
          previous[signal_name] = trap(signal_name) { stop_server(server) }
        rescue ArgumentError
          nil
        end
        yield
      ensure
        previous.each do |signal_name, handler|
          trap(signal_name, handler)
        rescue ArgumentError
          nil
        end
      end

      def stop_server(server)
        return unless server.respond_to?(:running?)
        return unless server.running?

        server.stop
      rescue StandardError
        nil
      end
    end

    class StdioServeBridge
      def initialize(server:, stdin: $stdin, stdout: $stdout, stderr: $stderr)
        @server = server
        @stdin = stdin.dup
        @stdout = stdout.dup
        @stderr = stderr
        @socket_dir = nil
        @socket_path = nil
        @socket = nil
        @stdin_thread = nil
        @stdout_thread = nil
        @closed = false
        @lock = Mutex.new

        @stdin.binmode if @stdin.respond_to?(:binmode)
        @stdout.binmode if @stdout.respond_to?(:binmode)
      end

      def start
        @socket_dir = Dir.mktmpdir("holons-stdio-serve-")
        @socket_path = File.join(@socket_dir, "bridge.sock")
      end

      def connect_to_server(timeout: 5.0)
        redirect_stdout_to_stderr

        deadline = monotonic_now + timeout
        begin
          @socket = UNIXSocket.new(@socket_path)
        rescue Errno::ENOENT, Errno::ECONNREFUSED
          raise "failed to connect to stdio bridge #{@socket_path}" if monotonic_now >= deadline

          sleep 0.05
          retry
        end

        @socket.binmode
        @stdin_thread = Thread.new { stdin_to_socket_loop }
        @stdout_thread = Thread.new { socket_to_stdout_loop }
      end

      def close
        socket = nil
        stdin = nil
        stdout = nil
        socket_path = nil
        socket_dir = nil

        @lock.synchronize do
          return if @closed

          @closed = true
          socket = @socket
          stdin = @stdin
          stdout = @stdout
          socket_path = @socket_path
          socket_dir = @socket_dir
          @socket = nil
          @stdin = nil
          @stdout = nil
          @socket_path = nil
          @socket_dir = nil
        end

        [socket, stdin, stdout].compact.each do |io|
          close_io(io)
        end

        [@stdin_thread, @stdout_thread].compact.each do |thread|
          thread.join(0.2)
        rescue StandardError
          nil
        end

        File.delete(socket_path) if socket_path && File.exist?(socket_path)
        Dir.rmdir(socket_dir) if socket_dir && Dir.exist?(socket_dir)
      rescue SystemCallError
        nil
      end

      private

      def redirect_stdout_to_stderr
        return if @stdout_redirected

        $stdout.flush if $stdout.respond_to?(:flush)
        @stderr.flush if @stderr.respond_to?(:flush)
        STDOUT.reopen(@stderr)
        STDOUT.sync = true
        @stdout_redirected = true
      rescue StandardError
        nil
      end

      def stdin_to_socket_loop
        loop do
          chunk = @stdin.readpartial(16 * 1024)
          @socket.write(chunk)
          @socket.flush if @socket.respond_to?(:flush)
        end
      rescue EOFError, IOError, SystemCallError
        nil
      ensure
        begin
          @socket.close_write unless @socket.nil? || @socket.closed?
        rescue IOError, SystemCallError
          nil
        end
        stop_server_once
      end

      def socket_to_stdout_loop
        loop do
          chunk = @socket.readpartial(16 * 1024)
          @stdout.write(chunk)
          @stdout.flush if @stdout.respond_to?(:flush)
        end
      rescue EOFError, IOError, SystemCallError
        nil
      ensure
        stop_server_once
      end

      def stop_server_once
        @lock.synchronize do
          return if @server_stopped

          @server_stopped = true
          @server.stop
        end
      rescue StandardError
        nil
      end

      def close_io(io)
        return if io.closed?

        io.close
      rescue IOError, SystemCallError
        nil
      end

      def monotonic_now
        Process.clock_gettime(Process::CLOCK_MONOTONIC)
      end
    end
  end
end

File: sdk/python-holons/holons/serve.py
from __future__ import annotations

"""Standard gRPC server runner for Python holons."""

import logging
import os
import shutil
import signal
import socket
import sys
import tempfile
import threading
import time
from concurrent import futures
from typing import Callable
from dataclasses import dataclass
from urllib.parse import urlparse

import grpc
from grpc_reflection.v1alpha import reflection

from holons import describe
from holons import grpcclient
from holons import observability
from holons.transport import DEFAULT_URI, scheme
from holons.v1 import observability_pb2, observability_pb2_grpc

logger = logging.getLogger("holons.serve")

RegisterFunc = Callable[[grpc.Server], None]
_MAX_GRPC_MESSAGE_BYTES = 1 << 20


@dataclass(frozen=True)
class ParsedFlags:
    listen_uri: str = DEFAULT_URI
    reflect: bool = False


@dataclass(frozen=True)
class MemberRef:
    slug: str
    address: str
    uid: str = ""


@dataclass(frozen=True)
class ChildSpec:
    slug: str
    binary: str


@dataclass(frozen=True)
class ServeOptions:
    reflect: bool = False
    member_endpoints: tuple[MemberRef, ...] = ()
    slug: str = ""


def parse_flags(args: list[str]) -> str:
    """Extract --listen or --port from command-line args."""
    return parse_options(args).listen_uri


def parse_options(args: list[str]) -> ParsedFlags:
    """Extract --listen, --port, and --reflect from command-line args."""
    listen_uri = DEFAULT_URI
    reflect_enabled = False

    for i, arg in enumerate(args):
        if arg == "--listen" and i + 1 < len(args):
            listen_uri = args[i + 1]
        if arg == "--port" and i + 1 < len(args):
            listen_uri = f"tcp://:{args[i + 1]}"
        if arg == "--reflect":
            reflect_enabled = True

    return ParsedFlags(listen_uri=listen_uri, reflect=reflect_enabled)


def parse_child_flags(args: list[str]) -> tuple[list[ChildSpec], list[str]]:
    """Extract repeated --child <slug>=<binary> flags from args."""
    children: list[ChildSpec] = []
    remaining: list[str] = []
    index = 0
    while index < len(args):
        arg = args[index]
        if arg == "--child" and index + 1 < len(args):
            child = _parse_child_spec(args[index + 1])
            if child is not None:
                children.append(child)
            index += 2
            continue
        if arg.startswith("--child="):
            child = _parse_child_spec(arg.removeprefix("--child="))
            if child is not None:
                children.append(child)
            index += 1
            continue
        remaining.append(arg)
        index += 1
    return children, remaining


def _parse_child_spec(raw: str) -> ChildSpec | None:
    if "=" not in raw:
        return None
    slug, binary = raw.split("=", 1)
    slug = slug.strip()
    binary = binary.strip()
    if not slug or not binary:
        return None
    return ChildSpec(slug=slug, binary=binary)


ParseChildFlags = parse_child_flags


def run(listen_uri: str, register_fn: RegisterFunc) -> None:
    """Start a gRPC server with reflection disabled by default."""
    run_with_options(listen_uri, register_fn, reflect=False)


def run_with_options(
    listen_uri: str,
    register_fn: RegisterFunc,
    reflect: bool = False,
    max_workers: int = 10,
    on_listen: Callable[[str], None] | None = None,
) -> None:
    run_with_serve_options(
        listen_uri,
        register_fn,
        ServeOptions(reflect=reflect),
        max_workers=max_workers,
        on_listen=on_listen,
    )


def run_with_serve_options(
    listen_uri: str,
    register_fn: RegisterFunc,
    options: ServeOptions | None = None,
    max_workers: int = 10,
    on_listen: Callable[[str], None] | None = None,
) -> None:
    """Start a gRPC server on the given transport URI.

    Native gRPC python transports: tcp://, unix://
    Bridged transports: stdio://
    """
    if options is None:
        options = ServeOptions()
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=max_workers),
        options=[
            ("grpc.max_receive_message_length", _MAX_GRPC_MESSAGE_BYTES),
            ("grpc.max_send_message_length", _MAX_GRPC_MESSAGE_BYTES),
        ],
    )
    observability.check_env()
    obs = None
    if os.environ.get("OP_OBS", "").strip():
        active = observability.current()
        obs = active if active.families else observability.from_env(observability.Config(slug=options.slug))
    register_fn(server)
    _register_holon_meta(server)
    if obs is not None and obs.families:
        observability.register_service(server, obs)

    reflection_enabled = False
    if options.reflect:
        service_names = [reflection.SERVICE_NAME]
        for handler in getattr(server._state, "generic_handlers", ()):
            service_name = getattr(handler, "service_name", None)
            if callable(service_name):
                name = service_name()
                if name and name not in service_names:
                    service_names.append(name)
        reflection.enable_server_reflection(tuple(service_names), server)
        reflection_enabled = True

    transport = scheme(listen_uri)
    stdio_bridge = None

    if transport == "tcp":
        addr = listen_uri[6:]
        port = server.add_insecure_port(addr)
        host = addr.rpartition(":")[0] or "0.0.0.0"
        actual_uri = f"tcp://{host}:{port}"
    elif transport == "unix":
        path = listen_uri[7:]
        server.add_insecure_port(f"unix:{path}")
        actual_uri = listen_uri
    elif transport == "stdio":
        # Bridge stdin/stdout to a temp Unix socket that grpcio can listen on
        stdio_bridge = _StdioServeBridge()
        sock_path = stdio_bridge.start()
        server.add_insecure_port(f"unix:{sock_path}")
        actual_uri = "stdio://"
    else:
        raise ValueError(
            "gRPC Python server supports tcp://, unix://, and stdio:// "
            f"in run_with_options(). For {transport}://, use holons.transport.listen() "
            "with a custom server loop."
        )

    mode = "reflection ON" if reflection_enabled else "reflection OFF"
    prom_server = _start_prom_server(obs)
    metrics_addr = prom_server[1] if prom_server is not None else ""
    if obs is not None and obs.families:
        _start_observability_runtime(obs, actual_uri, transport, metrics_addr)
    if on_listen is not None:
        on_listen(actual_uri)
    logger.info("gRPC server listening on %s (%s)", actual_uri, mode)

    server.start()
    started_relays = _start_member_relays(obs, options.member_endpoints)

    # For stdio, start bridging after the server is ready
    if stdio_bridge is not None:
        stdio_bridge.connect_to_server()

    def _shutdown(*_args):
        logger.info("shutting down gRPC server")
        server.stop(10)

    signal.signal(signal.SIGTERM, _shutdown)
    signal.signal(signal.SIGINT, _shutdown)

    print(f"gRPC server listening on {actual_uri} ({mode})", file=sys.stderr)

    try:
        server.wait_for_termination()
    finally:
        for relay, channel in started_relays:
            relay.stop()
            _close_channel(channel)
        if prom_server is not None:
            prom_server[0].close()
        if stdio_bridge is not None:
            stdio_bridge.close()


def _start_observability_runtime(
    obs: observability.Observability,
    actual_uri: str,
    transport: str,
    metrics_addr: str = "",
) -> None:
    if not obs.cfg.run_dir:
        return
    observability.enable_disk_writers(obs.cfg.run_dir)
    if obs.enabled(observability.Family.EVENTS):
        obs.emit(
            observability.EventType.INSTANCE_READY,
            {"listener": actual_uri, "metrics_addr": metrics_addr},
        )
    observability.write_meta_json(
        obs.cfg.run_dir,
        observability.MetaJSON(
            slug=obs.cfg.slug,
            uid=obs.cfg.instance_uid,
            pid=os.getpid(),
            started_at=time.time(),
            transport=transport,
            address=actual_uri,
            metrics_addr=metrics_addr,
            log_path=os.path.join(obs.cfg.run_dir, "stdout.log") if obs.enabled(observability.Family.LOGS) else "",
            organism_uid=obs.cfg.organism_uid,
            organism_slug=obs.cfg.organism_slug,
        ),
    )


def _register_holon_meta(server: grpc.Server) -> None:
    try:
        describe.register(server)
    except describe.IncodeDescriptionError as exc:
        logger.error("HolonMeta registration failed: %s", exc)
        raise
    except Exception as exc:
        logger.exception("HolonMeta registration failed: %s", exc)
        raise RuntimeError(f"register HolonMeta: {exc}") from exc


def _start_prom_server(
    obs: observability.Observability | None,
) -> tuple[observability.PromServer, str] | None:
    if obs is None or not obs.enabled(observability.Family.PROM):
        return None
    addr = obs.cfg.prom_addr or ":0"
    server = observability.PromServer(addr)
    try:
        metrics_addr = server.start()
    except Exception as exc:
        logger.warning("warning: prom HTTP bind failed: %s", exc)
        return None
    logger.info("Prometheus /metrics listening on %s", metrics_addr)
    return server, metrics_addr


def _start_member_relays(
    obs: observability.Observability | None,
    members: tuple[MemberRef, ...],
) -> list[tuple[observability.MemberRelay, grpc.Channel]]:
    if (
        obs is None
        or (
            not obs.enabled(observability.Family.LOGS)
            and not obs.enabled(observability.Family.EVENTS)
        )
    ):
        return []
    started: list[tuple[observability.MemberRelay, grpc.Channel]] = []
    for raw in members:
        member = MemberRef(
            slug=(raw.slug or "").strip(),
            uid=(raw.uid or "").strip(),
            address=(raw.address or "").strip(),
        )
        if not member.slug or not member.address:
            logger.warning(
                'warning: observability relay skipped incomplete member ref: slug="%s" uid="%s" address="%s"',
                member.slug,
                member.uid,
                member.address,
            )
            continue
        channel = None
        try:
            channel = grpcclient.dial_uri(_normalize_relay_dial_target(member.address))
            member = _resolve_relay_member_identity(channel, member)
            if not member.uid:
                logger.warning(
                    "warning: observability relay uid unresolved for %s at %s; chain hops will have empty uid",
                    member.slug,
                    member.address,
                )
            relay = observability.MemberRelay(
                child_slug=member.slug,
                child_uid=member.uid,
                channel=channel,
                observability=obs,
            )
            relay.start()
            started.append((relay, channel))
            channel = None
        except Exception as exc:
            if channel is not None:
                _close_channel(channel)
            logger.warning(
                "warning: observability relay start %s/%s: %s",
                member.slug,
                member.uid,
                exc,
            )
    return started


def _resolve_relay_member_identity(channel: grpc.Channel, member: MemberRef) -> MemberRef:
    if member.uid:
        return member
    client = observability_pb2_grpc.HolonObservabilityStub(channel)
    try:
        stream = client.Events(
            observability_pb2.EventsRequest(
                types=[observability_pb2.INSTANCE_READY],
                follow=False,
            ),
            timeout=2.0,
        )
        for event in stream:
            if not event.instance_uid or event.chain:
                continue
            slug = event.slug.strip() or member.slug
            return MemberRef(slug=slug, uid=event.instance_uid.strip(), address=member.address)
    except Exception:
        pass

    try:
        snap = client.Metrics(observability_pb2.MetricsRequest(), timeout=2.0)
        if snap.instance_uid:
            slug = snap.slug.strip() or member.slug
            return MemberRef(slug=slug, uid=snap.instance_uid.strip(), address=member.address)
    except Exception:
        pass
    return member


def _normalize_relay_dial_target(target: str) -> str:
    trimmed = target.strip()
    if "://" not in trimmed:
        return trimmed
    parsed = urlparse(trimmed)
    if parsed.scheme != "tcp":
        return trimmed
    host = parsed.hostname or "127.0.0.1"
    if host in ("0.0.0.0", "::"):
        host = "127.0.0.1"
    if parsed.port is None:
        return trimmed
    return f"tcp://{host}:{parsed.port}"


def _close_channel(channel: grpc.Channel) -> None:
    close = getattr(channel, "close", None)
    if callable(close):
        close()


class _StdioServeBridge:
    """Bridge stdin/stdout to a Unix socket for grpcio server-side stdio.

    grpcio's Python binding cannot accept raw file descriptors, so we
    create a temporary Unix socket, make grpcio listen on it, then
    forward bytes between stdin/stdout and the socket.
    """

    def __init__(self) -> None:
        self._sock_dir: str | None = None
        self._sock_path: str | None = None
        self._conn: socket.socket | None = None
        self._closed = threading.Event()
        self._stdin_fd = os.dup(0)
        self._stdout_fd = os.dup(1)
        # Redirect stdout to stderr so print() doesn't corrupt the data channel
        os.dup2(2, 1)

    def start(self) -> str:
        """Create the Unix socket path (does not connect yet)."""
        self._sock_dir = tempfile.mkdtemp(prefix="holons-stdio-serve-")
        self._sock_path = os.path.join(self._sock_dir, "bridge.sock")
        return self._sock_path

    def connect_to_server(self) -> None:
        """Connect to the grpcio Unix socket and start bridging threads."""
        assert self._sock_path is not None

        # Retry connecting — grpcio may not be listening yet
        conn = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        for attempt in range(20):
            try:
                conn.connect(self._sock_path)
                break
            except (ConnectionRefusedError, FileNotFoundError):
                import time
                time.sleep(0.05)
        else:
            raise RuntimeError("failed to connect to grpcio Unix socket")
        self._conn = conn

        # stdin → socket (client data arriving on stdin goes to grpc server)
        t1 = threading.Thread(target=self._stdin_to_socket, daemon=True)
        t1.start()

        # socket → stdout (grpc server responses go out on stdout)
        t2 = threading.Thread(target=self._socket_to_stdout, daemon=True)
        t2.start()

    def close(self) -> None:
        if self._closed.is_set():
            return
        self._closed.set()

        conn = self._conn
        self._conn = None
        if conn is not None:
            try:
                conn.shutdown(socket.SHUT_RDWR)
            except OSError:
                pass
            try:
                conn.close()
            except OSError:
                pass

        for fd in (self._stdin_fd, self._stdout_fd):
            try:
                os.close(fd)
            except OSError:
                pass

        if self._sock_path:
            try:
                os.unlink(self._sock_path)
            except OSError:
                pass
        if self._sock_dir:
            shutil.rmtree(self._sock_dir, ignore_errors=True)

    def _stdin_to_socket(self) -> None:
        """Forward bytes from stdin to the Unix socket."""
        conn = self._conn
        assert conn is not None

        while not self._closed.is_set():
            try:
                data = os.read(self._stdin_fd, 64 * 1024)
            except OSError:
                break
            if not data:
                break
            try:
                conn.sendall(data)
            except OSError:
                break

        # EOF on stdin — shut down the write side of the socket
        try:
            conn.shutdown(socket.SHUT_WR)
        except OSError:
            pass

        # Stdin closed means the parent process disconnected; trigger shutdown
        os.kill(os.getpid(), signal.SIGTERM)

    def _socket_to_stdout(self) -> None:
        """Forward bytes from the Unix socket to stdout."""
        conn = self._conn
        assert conn is not None

        while not self._closed.is_set():
            try:
                data = conn.recv(64 * 1024)
            except OSError:
                break
            if not data:
                break
            try:
                os.write(self._stdout_fd, data)
            except OSError:
                break
